### PR TITLE
New Feature: add DropBlock layer

### PR DESCRIPTION
**Pull request by xiaohu2015**
As discussed in https://github.com/pytorch/vision/issues/5414, I am working to add DropBlock layer to Torchvision.

we add DropBlock2d and DropBlock3d module.

### Reference
- https://github.com/tensorflow/tpu/blob/master/models/official/resnet/resnet_model.py#L276
- https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/layers/drop.py
- https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/plugins/dropblock.py

### CODE REVIEW

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:20]: Nit:
```suggestion
        super().__init__()
```

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:37]: Skip early if p is zero:
```suggestion
        if not self.training or self.p == 0.0:
```

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:53]: Perhaps the following will be faster?
```suggestion
            input.mul_(mask).mul_(normalize_scale)
```

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:46]: Could you provide a reference for the choice of `mask_shape` and the subsequent padding with zeros that you do? I need to check the paper in more detail but my expectation was that we would sample from bernoulli with shape equal the input. I'll need to have a closer look on the paper to remember the details.

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:51]: Shall we make `eps=1e-6` configurable on the constructor of the class?

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:29]: I recommend moving the entire code of `forward` to a separate public method so that it's available on a functional API. See the `stochastic_depth` method for [reference](https://github.com/pytorch/vision/blob/main/torchvision/ops/stochastic_depth.py#L8).

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:44]: Worth refactoring this to use the following idiom:
https://github.com/pytorch/vision/blob/93c85bbcc31f8d5a052daf06f2f91f39697af1a4/torchvision/ops/stochastic_depth.py#L40-L41

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:8]: Creating new tensors is not supported by FX, so we will need to declare it:
https://github.com/pytorch/vision/blob/93c85bbcc31f8d5a052daf06f2f91f39697af1a4/torchvision/ops/stochastic_depth.py#L47

File: torchvision/ops/dropblock.py
Comment (datumbox) [inline torchvision/ops/dropblock.py:60]: There are a few boilerplate things that need to be done when adding an operator such as adding it to docs, including it on init files, adding tests etc etc. Here is an equivalent PR that you can follow as an example: #4301

**Review by datumbox — COMMENTED** [review]
@xiaohu2015 Awesome work as always. Thanks a lot!

The implementation looks very nice overall. I've added a couple of comments concerning nits and boilerplate code needed for ops in TorchVision. My only real question is on the way you do the sampling on bernoulli, see below.

Finally, would you be interested in implementing the 3d version on the same PR? We want to start offering better video support and this contribution will help us offer better models. If you prefer, we can also do it on a follow up.

File: torchvision/ops/dropblock.py
Comment (xiaohu2015) [reply torchvision/ops/dropblock.py:46]: @datumbox the implementation is consistent with the original paper, there is 3 steps:
1. generate the center mask, which means the block center, so the center mask shape : `mask_shape = (N, C, H - self.block_size + 1, W - self.block_size + 1)`; we also should convert the `drop_prob` to `gamma` of the bernoulli.
2. convert the center mask to block mask: `pad` and use a `maxpool` to get block mask.
3. use the block mask to mask the input.

Comment (datumbox) [reply torchvision/ops/dropblock.py]: I've checked the paper in detail and you are right. Page 3, figure 2a shows that the center mask should be drawn from the green area, hence your approach is correct. Also this implementation aligns closely with the [original TF implementation](https://github.com/tensorflow/tpu/blob/a6005716b311f96190590414479f0261726270e7/models/official/resnet/resnet_model.py#L315-L316).

**Comment by datumbox** [comment]
@xiaohu2015 I've checked the paper and the original TF implementation and I believe your approach looks great. Thanks a lot for the contribution.

The only pendings are:
- Fix tests (explicitly `import torch.fx` to solve some of the [failing tests] + fix the [linter](https://app.circleci.com/pipelines/github/pytorch/vision/14924/workflows/86e431cb-a503-412a-8d13-0424bb7ea59b/jobs/1201333))
- Add the new ops on `docs/source/ops.rst`
- Add a test on `test/test_ops.py` (see #4301)

**Comment by xiaohu2015** [comment]
> @xiaohu2015 I've checked the paper and the original TF implementation and I believe your approach looks great. Thanks a lot for the contribution.
> 
> The only pendings are:
> 
> * Fix tests (explicitly `import torch.fx` to solve some of the [failing tests](https://app.circleci.com/pipelines/github/pytorch/vision/14924/workflows/c4da8b86-f1ae-4809-8991-01ea8ad4274e/jobs/1201344) + fix the [linter](https://app.circleci.com/pipelines/github/pytorch/vision/14924/workflows/86e431cb-a503-412a-8d13-0424bb7ea59b/jobs/1201333))
> * Add the new ops on `docs/source/ops.rst`
> * Add a test on `test/test_ops.py` (see [Add StochasticDepth implementation #4301](https://github.com/pytorch/vision/pull/4301))

OK

File: torchvision/ops/drop_block.py
Comment (xiaohu2015) [reply torchvision/ops/drop_block.py]: oh, I remove this.

File: test/test_ops.py
Comment (datumbox) [inline test/test_ops.py:1399]: Is this because it's either going to have selected 1 center and thus the whole block will turn to zeros OR no center in which case no zeros?

File: test/test_ops.py
Comment (datumbox) [inline test/test_ops.py:1377]: The testing strategy here is not great as we only validate the input for a fraction of the inputs. An alternative approach would be to try to count the zeroes based on the sampling probability. As noted on the paper, the number of zeroes will be approximate due to overlaps but perhaps we can create a not so flaky test for this.

If you prefer, we can do this on a follow up PR.

**Review by datumbox — COMMENTED** [review]
@xiaohu2015 The changes look good to me. I've pushed to your branch to fix one remaining linter issue. Let's see if the CI passes now.

File: test/test_ops.py
Comment (xiaohu2015) [reply test/test_ops.py:1399]: yes

File: test/test_ops.py
Comment (xiaohu2015) [reply test/test_ops.py:1377]: I have read the test function of  https://github.com/pytorch/vision/pull/4301, is it necessary check this? I found there is no such check for `nn.Dropout` in PyTorch.

Comment (datumbox) [reply test/test_ops.py:1377]: Yes that's what I had in mind. The test of `StochasticDepth` confirms that the number of drops are within the expended confidence intervals. These intervals should be larger at DropBlock because of the overlapping blocks.

Concerning whether this test is necessary, the answer is I wouldn't block the PR for this but it will protect us from someone breaking the op accidentally on the future. I'm surprised you say that `Dropout` is not properly checked at PyTorch Core. Perhaps the test is not on an obvious location (I know dropout has tests in a few different places)?

Comment (xiaohu2015) [reply test/test_ops.py:1377]: @datumbox oh, maybe I missed some codes.

**Comment by datumbox** [comment]
@xiaohu2015 All tests pass. 🚀 🚀 

On the positive side, your implementation seems correct to me, fully aligns with the paper. Though some of the implementation details differ from the references you provided, after doing a deep dive I believe that your approach is equivalent. Some limited testing also shows that the implementation works as intended.

On the negative side, the behaviour of the method is not fully tested (see points at  https://github.com/pytorch/vision/pull/5416#discussion_r811818770). Moreover, we haven't trained a model to show-case that using the layer leads to the expected increase.

What is your bandwidth at this point? Do you have the capacity to do the above 2 checks prior merging? Do you think it's something you would do on a follow up PR after merge? It's a bit hard to provide support on this one due to limited bandwidth on our side because of the upcoming release and other work on the pipeline. Another approach could be to ping a couple of other regular contributors to see if they can help us test it. Let me know your preference/thoughts. Thanks!

File: test/test_ops.py
Comment (xiaohu2015) [reply test/test_ops.py:1377]: @datumbox I add the random test check. But, just as you said, the actual sampling probability is slightly lower than the setting value because of overlaps in dropblock.

File: test/test_ops.py
Comment (datumbox) [inline test/test_ops.py:1375]: Do we need all the combinations of `p` and `block_sizes` given we only assert to a handful of experiments? 

Instead of doing a cartesian product of p x block_sizes, we might want to test just a couple of specific of them to speed the tests.

**Review by datumbox — APPROVED** [review]
LGTM, @xiaohu2015.

I've added one last optional test improvement for your consideration below. The testing strategy looks good to me. Let me know when the training checks is completed to merge. 

Thanks a lot as always for the awesome contribution!

File: test/test_ops.py
Comment (xiaohu2015) [reply test/test_ops.py:1375]: OK, I remove some.

**Comment by xiaohu2015** [comment]
> LGTM, @xiaohu2015.
> 
> I've added one last optional test improvement for your consideration below. The testing strategy looks good to me. Let me know when the training checks is completed to merge.
> 
> Thanks a lot as always for the awesome contribution!

I just run 90-epoch Resnet50, maybe I can get result next day.

**Comment by datumbox** [comment]
@xiaohu2015 Awesome. The paper uses 90 epochs for the baseline and 250 epochs for those with Dropblock (see page 4, section 4.1 on [paper](https://arxiv.org/pdf/1810.12890.pdf)). If you don't have enough compute to run this, ping me on slack and we can work something out.

FYI the last update broke the [linter](https://app.circleci.com/pipelines/github/pytorch/vision/14999/workflows/985b2195-ebf7-41ef-b3c1-b85af9677da3/jobs/1208564).

**Comment by datumbox** [comment]
Here is my branch for testing the ResNet50 with DropBlock: #5483 

We are not going to used scheduled DropBlock. I'll let you know about the results once I got them.

**Comment by datumbox** [comment]
@xiaohu2015 I got good news. The addition of your DropBlock seems to improve the accuracy. 

Here is the baseline ResNet50 accuracy for a 90-epoch training without DropBlock:
`Test:  Acc@1 76.130 Acc@5 92.862`

Here is the one if we add DropBlock on Groups 3&4 wo/ scheduling (implementation details [here](https://github.com/datumbox/vision/blob/4e76a423bc0ef8c7065125f60745f4745da44885/torchvision/models/resnet.py)):
`Test:  Acc@1 77.064 Acc@5 93.436`

Both the baseline and DropBlock values are different from the ones reported on the paper (baseline 76.51 and DropBlock ~77.5), but I think that's the result of other minor differences that we didn't replicate (hyper-params, small architectural improvements etc). I think we have sufficient evidence that your implementation work as expected and we can merge the PR. Thanks so much for the awesome contribution and work!

**Pull request merged**
